### ISSUE TITLE
Why are local functions sometimes unnecessarily emited as internal?

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#23303]
Local functions are logically private, so I think the emitted IL should reflect that, but it usually doesn't.

Here is what the emitted code currently looks like, depending on the closure of the local function:

1. [If the local function has no closure, it is emitted as `internal`, when it could be `private`:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMACAwgWAFADe+GJGKALBgLIAUAlMaUXqa2ZQDYD2AxhB/QwEAvoxKi8woA=)

    ```c#
    void M()
    {
        void local() {}
    }
    ```

    ```c#
    [CompilerGenerated]
    internal static void <M>g__local|0_0()
    {
    }
    ```

2. [If the local function closes over `this`, it is correctly emitted as `private`:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMACAwgWAFADe+GJGKALBgLIAUAlMaUXqa9hgDYD2AxhB/QwBeAHwYYACwCWUANyMSAX3yKgA===)

    ```c#
    void M()
    {
        C local() => this;
    }
    ```

    ```c#
    [CompilerGenerated]
    private C <M>g__local|0_0()
    {
        return this;
    }
    ```

3. [If the local function closes over a local variable and the closure is a `struct` passed by `ref`, it is emitted as `internal` when it could be `private`:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMACAwgWAFADe+GJGKALBgLIAUAlMaUXqaxgJYB2MHGAvBnJoA3IzYduGADYB7AMYQp9fgD4OolqQC++LUA=)

    ```c#
    void M()
    {
        int i = 42;
        int local() => i;
    }
    ```

    ```c#
    [CompilerGenerated]
    [StructLayout(LayoutKind.Auto)]
    private struct <>c__DisplayClass0_0
    {
        public int i;
    }

    [CompilerGenerated]
    internal static int <M>g__local|0_0(ref C.<>c__DisplayClass0_0 ptr)
    {
        return ptr.i;
    }
    ```

4. [If the local function closes over a local variable and the closure is a `class`, it is correctly emitted as `internal` method on the closure:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgiUAmDAYSIG8iN6sAWDAWQAoBKOh2wh/jAEsAdjCEYAvBkakCfAfRQBWADwiYAPgwAbCGGAATCJIydJWwXIX11OgPYBjCNrMSLVhgF8inoA)

    ```c#
    void M()
    {
        int i = 42;
        Func<int> lambda = () => i;
        int local() => i;
    }
    ```

    ```c#
    [CompilerGenerated]
    private sealed class <>c__DisplayClass0_0
    {
        public int i;

        internal int <M>b__0()
        {
            return this.i;
        }

        internal int <M>g__local|1()
        {
            return this.i;
        }
    }
    ```

In case 4, the emitted method is correctly `internal`, because it is a member of the (`private`) closure class. But in cases 1 and 3, the emitted method could be `private`, but it's `internal`.

I believe the code that causes this behavior is:

https://github.com/dotnet/roslyn/blob/c505c8faacb0bfb48a15720f7b423725bbdc9795/src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureMethod.cs#L106

**Comment by huoyaoyuan** [comment]
The code you're showing are generating `internal static` instead of `private static`.
I think the modifier should be determined by in which class to generate the function. `internal` should be only used when generating a closure class.

**Comment by jcouv** [comment]
Tagging @agocke 
Btw, Andy, I looked in roslyn and csharplang repos for documentation on local functions and found little more than grammar.

**Comment by agocke** [comment]
Yup, this could be cleaned up. Not a high priority -- these methods are unspeakable (can't be called by C# code). To a certain extent, the accessibility doesn't matter at all.

**Comment by jcouv** [comment]
Moved to 16.0 based on priority. Feel free to bring back if you have time to address.

**Comment by jcouv** [comment]
Moved to Compiler.Next milestone
